Fix for internal side indicator on displaced mesh

GiudGiud opened the pull request:
see #30015 and discussion https://github.com/idaholab/moose/discussions/29699

GiudGiud committed:
dont rely on nodal index for non-nodal variables in the internal side indicator
ban displaced mesh in indicator (can be used outside) since this leads to a missing (uninitialized) qrule for nodal variables
closes #30015

GiudGiud committed:
Add test for AMR + displaced mesh + internal jump indicator refs #30015

GiudGiud committed:
Add a test for using const monomial on a displaced mesh
Robustify block restriction + adaptivity test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2727975/), step Clang format on 7fcfd38 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30005/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30005/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e6f742bded1f28a5d4fa5991eff1b7008b3a7d2d`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2737013/), step Generate coverage on 80ed3f4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2aabd6</th>
      <th colspan=4>#30005 80ed3f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.26%</td>
      <td>85.26%</td>
      <td>-0.00%</td>
      <td>92.86%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108581</td>
      <td>108588</td>
      <td>+7</td>
      <td>13</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18769</td>
      <td>18772</td>
      <td>+3</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30005/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30005/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30005/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
@lindsayad @roystgnr would you have time to review?

GiudGiud committed:
Address Alex's review:
- remove unused attributed
- ban non-const monomials from being anything else than const monomials
- dont lock the solution vector, trust libmesh to lock
- better requirement

GiudGiud commented on framework/src/indicators/InternalSideIndicator.C:
```diff
@@ -74,29 +74,39 @@ InternalSideIndicator::InternalSideIndicator(const InputParameters & parameters)
-    _solution.add(_field_var.nodalDofIndexNeighbor(), sum * _neighbor_elem->hmax());
+    const auto sys_num = _field_var.sys().number();
+    const auto var_num = _field_var.number();
+    _solution.add(_current_elem->dof_number(sys_num, var_num, 0), sum * _current_elem->hmax());
+    if (_field_var.hasBlocks(_neighbor_elem->subdomain_id()))
+      _solution.add(_neighbor_elem->dof_number(sys_num, var_num, 0), sum * _neighbor_elem->hmax());
```
done

GiudGiud commented on framework/src/indicators/InternalSideIndicator.C:
```diff
@@ -74,29 +74,39 @@ InternalSideIndicator::InternalSideIndicator(const InputParameters & parameters)
 
+  // Contribution is added to the two elements
   {
     Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
```
done

GiudGiud commented on test/tests/adaptivity/displaced_mesh/tests:
```diff
@@ -0,0 +1,28 @@
+                AuxVariables/disp_x/family=LAGRANGE AuxVariables/disp_y/family=LAGRANGE
+                Adaptivity/Indicators/indicator/use_displaced_mesh=false
+                Outputs/hide='disp_x disp_y'"
+    requirement = "The system shall allow for using internal side indicators on the displaced mesh."
```
changed a little

GiudGiud replied:
the displaced mesh part and the monomial part are different

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2804557/), step Clang format on c8c82b0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30005/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30005/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d96843ae72a88762155d04ba6095421ba4bddd70`

lindsayad commented:
I approve the spirit of the PR 😆 I trust @GiudGiud to fix the failures before merging
